### PR TITLE
Update control owning field of MonthCalendar to using weak reference

### PR DESCRIPTION
**Pull request by SimonZhao888**
Related to https://github.com/dotnet/winforms/issues/9224

## Proposed changes

- Update the control owning field of MonthCalendar to using weak references
- Enable the GC collect test for the MonthCalendar control in unit test cases

## Customer Impact

- The strong reference used by the original Owner property can cause a memory leak

## Regression? 

- No

## Test methodology 

- Unit tests

## Test environment(s) 

- dotnet 8.0.0-preview.7.23358.3

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9454)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 1]
```diff
@@ -201,13 +201,14 @@ internal override int ColumnCount
             };
         }
 
-        internal DayOfWeek FirstDayOfWeek => CastDayToDayOfWeek(_owningMonthCalendar.FirstDayOfWeek);
+        internal DayOfWeek FirstDayOfWeek => this.TryGetOwnerAs(out MonthCalendar? owner) ? CastDayToDayOfWeek(owner.FirstDayOfWeek) : CastDayToDayOfWeek(Day.Default);
 
-        internal bool Focused => _owningMonthCalendar.Focused;
+        internal bool Focused => this.TryGetOwnerAs(out MonthCalendar? owner) ? owner.Focused : false;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:206]: Please unify local variable names. Either use `owner` or `owningMonthCalendar` everywhere

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 2]
```diff
@@ -26,9 +25,10 @@ internal class MonthCalendarAccessibleObject : ControlAccessibleObject
 
         public MonthCalendarAccessibleObject(MonthCalendar owner) : base(owner)
         {
-            _owningMonthCalendar = owner;
-
-            _owningMonthCalendar.DisplayRangeChanged += OnMonthCalendarStateChanged;
+            if(this.TryGetOwnerAs(out MonthCalendar? owningMonthCalendar))
+            {
+                owningMonthCalendar.DisplayRangeChanged += OnMonthCalendarStateChanged;
+            }
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:31]: Can we do it this way?
```suggestion
            owner.DisplayRangeChanged += OnMonthCalendarStateChanged;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 3]
```diff
@@ -201,13 +201,14 @@ internal override int ColumnCount
...
 
-        internal bool Focused => _owningMonthCalendar.Focused;
+        internal bool Focused => this.TryGetOwnerAs(out MonthCalendar? owningMonthCalendar) ? owningMonthCalendar.Focused : false;
 
         internal CalendarCellAccessibleObject? FocusedCell
             => UiaCore.UiaClientsAreListening()
-                ? _focusedCellAccessibleObject ??= GetCellByDate(_owningMonthCalendar._focusedDate)
+                ? _focusedCellAccessibleObject ??= this.TryGetOwnerAs(out MonthCalendar? owningMonthCalendar) ? GetCellByDate(owningMonthCalendar._focusedDate)
+                : null
                 : null;
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:212]: It's not related to this PR, but could you also remove `UiaCore.UiaClientsAreListening()` check here? It's might have been needed for `FocusedCell?.RaiseAutomationEvent(...)` usages, but `RaiseAutomationEvent` already has this check:
https://github.com/dotnet/winforms/blob/b5695a8d2321deebc8885610b6fb8ced4d7933e3/src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.cs#L2350-L2359

We also don't use this check in any other AO, so this seems redundant.

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 4]
```diff
@@ -569,20 +568,37 @@ internal CalendarTodayLinkAccessibleObject TodayLinkAccessibleObject
...
-                        return $"{_owningMonthCalendar.SelectionStart:y}";
+                        if(this.TryGetOwnerAs(out owningMonthCalendar))
+                        {
+                            return $"{owningMonthCalendar.SelectionStart:y}";
+                        }
+                        else
+                        {
+                            return string.Empty;
+                        }
+
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:591]: Nit: `else` is not needed here and in similar places below.
```suggestion
                        if (this.TryGetOwnerAs(out owningMonthCalendar))
                        {
                            return $"{owningMonthCalendar.SelectionStart:y}";
                        }
                        
                        return string.Empty;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs [hunk 5]
```diff
@@ -569,20 +568,37 @@ internal CalendarTodayLinkAccessibleObject TodayLinkAccessibleObject
...
             get
             {
+                MonthCalendar? owningMonthCalendar;
                 SelectionRange? range;
 
                 switch (CalendarView)
                 {
                     case MONTH_CALDENDAR_MESSAGES_VIEW.MCMV_MONTH:
-                        range = _owningMonthCalendar.SelectionRange;
+                        range = this.TryGetOwnerAs(out owningMonthCalendar) ? owningMonthCalendar.SelectionRange : new SelectionRange();
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/MonthCalendar.MonthCalendarAccessibleObject.cs:577]: Perhaps we could return an empty string if owner is null? Now we return a range from `DateTime.MinValue` to `DateTime.MaxValue` (default values with parameter-less constructor) - not sure if it's a good value.

**Review by Tanya-Solyanik — APPROVED** [review]
Thank you!

**Pull request merged**